File: src/guidance/sync/json_writer.zig
//! sync/json_writer.zig — JSON serialization for guidance documents.
//!
//! Moved from types.zig to keep type definitions separate from serialization logic.
const std = @import("std");
const types = @import("../types.zig");
const common = @import("common");

// ---------------------------------------------------------------------------
// Primitives
// ---------------------------------------------------------------------------

fn writeEscaped(writer: anytype, value: []const u8) !void {
    try common.jsonWriteEscaped(writer, value);
}

/// Write `"key": "value"` — no leading whitespace, no trailing comma or newline.
fn writeStrField(writer: anytype, key: []const u8, value: []const u8) !void {
    try writer.print("\"{s}\": \"", .{key});
    try writeEscaped(writer, value);
    try writer.writeByte('"');
}

/// Emit `,\n` when `need.*` is already true, then set `need.*`.
/// Call before every field; the first call is a no-op that arms subsequent ones.
fn writeSep(writer: anytype, need: *bool) !void {
    if (need.*) try writer.writeAll(",\n");
    need.* = true;
}

/// Write `"key": ["a", "b"]` — no leading whitespace, no trailing comma or newline.
fn writeStringArray(writer: anytype, key: []const u8, items: []const []const u8) !void {
    try writer.print("\"{s}\": [", .{key});
    for (items, 0..) |item, i| {
        if (i > 0) try writer.writeAll(", ");
        try writer.writeByte('"');
        try writeEscaped(writer, item);
        try writer.writeByte('"');
    }
    try writer.writeByte(']');
}

// ---------------------------------------------------------------------------
// Member serialization
// ---------------------------------------------------------------------------

/// Write one Member as a JSON object directly to `writer`.
/// `indent` is prepended to the outer braces; field lines get `indent ++ "  "`.
/// No trailing newline or comma is written — the caller decides.
fn writeMember(writer: anytype, member: types.Member, indent: []const u8) !void {
    var fi_buf: [64]u8 = undefined;
    const fi = std.fmt.bufPrint(&fi_buf, "{s}  ", .{indent}) catch return error.IndentTooDeep;

    try writer.print("{s}{{\n", .{indent});

    var need: bool = false;

    try writeSep(writer, &need);
    try writer.print("{s}\"type\": \"{s}\"", .{ fi, @tagName(member.type) });

    try writeSep(writer, &need);
    try writer.print("{s}", .{fi});
    try writeStrField(writer, "name", member.name);

    if (member.match_hash) |h| {
        try writeSep(writer, &need);
        try writer.print("{s}", .{fi});
        try writeStrField(writer, "match_hash", h);
    }
    if (member.signature) |s| {
        try writeSep(writer, &need);
        try writer.print("{s}", .{fi});
        try writeStrField(writer, "signature", s);
    }
    if (member.params.len > 0) {
        try writeSep(writer, &need);
        try writer.print("{s}\"params\": [\n", .{fi});
        for (member.params, 0..) |param, i| {
            try writer.print("{s}  {{ \"name\": \"", .{fi});
            try writeEscaped(writer, param.name);
            try writer.writeByte('"');
            if (param.type) |t| {
                try writer.writeAll(", \"type\": \"");
                try writeEscaped(writer, t);
                try writer.writeByte('"');
            }
            if (param.default) |d| {
                try writer.writeAll(", \"default\": \"");
                try writeEscaped(writer, d);
                try writer.writeByte('"');
            }
            try writer.writeAll(" }");
            if (i < member.params.len - 1) try writer.writeByte(',');
            try writer.writeByte('\n');
        }
        try writer.print("{s}]", .{fi});
    }
    if (member.returns) |r| {
        try writeSep(writer, &need);
        try writer.print("{s}", .{fi});
        try writeStrField(writer, "returns", r);
    }
    if (member.tags.len > 0) {
        try writeSep(writer, &need);
        try writer.print("{s}", .{fi});
        try writeStringArray(writer, "tags", member.tags);
    }
    if (member.patterns.len > 0) {
        try writeSep(writer, &need);
        try writer.print("{s}\"patterns\": [\n", .{fi});
        for (member.patterns, 0..) |pat, i| {
            try writer.print("{s}  {{ \"name\": \"", .{fi});
            try writeEscaped(writer, pat.name);
            try writer.print("\", \"type\": \"{s}\"", .{@tagName(pat.type)});
            if (pat.ref) |ref| {
                try writer.writeAll(", \"ref\": \"");
                try writeEscaped(writer, ref);
                try writer.writeByte('"');
            }
            try writer.writeAll(" }");
            if (i < member.patterns.len - 1) try writer.writeByte(',');
            try writer.writeByte('\n');
        }
        try writer.print("{s}]", .{fi});
    }

    // is_pub is always present.
    try writeSep(writer, &need);
    try writer.print("{s}\"is_pub\": {any}", .{ fi, member.is_pub });

    if (member.members.len > 0) {
        var ni_buf: [64]u8 = undefined;
        const ni = std.fmt.bufPrint(&ni_buf, "{s}    ", .{indent}) catch return error.IndentTooDeep;

        try writeSep(writer, &need);
        try writer.print("{s}\"members\": [\n", .{fi});
        for (member.members, 0..) |nested, i| {
            try writeMember(writer, nested, ni);
            if (i < member.members.len - 1) try writer.writeByte(',');
            try writer.writeByte('\n');
        }
        try writer.print("{s}]", .{fi});
    }

    if (member.line) |l| {
        try writeSep(writer, &need);
        try writer.print("{s}\"line\": {d}", .{ fi, l });
    }

    try writer.print("\n{s}}}", .{indent});
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/// Serialize a Member to an owned byte slice.  The caller must free the result.
/// Returns null only on allocation failure; the ?[]u8 type is preserved for
/// backward compatibility with the re-export in types.zig.
pub fn jsonifyMember(allocator: std.mem.Allocator, member: types.Member) !?[]u8 {
    var list_aw: std.Io.Writer.Allocating = .init(allocator);
    errdefer list_aw.deinit();
    const writer = &list_aw.writer;
    try writeMember(writer, member, "");
    return try list_aw.toOwnedSlice();
}

pub fn jsonifyGuidanceDoc(allocator: std.mem.Allocator, doc: types.GuidanceDoc) ![]u8 {
    var list_aw: std.Io.Writer.Allocating = .init(allocator);
    errdefer list_aw.deinit();
    const writer = &list_aw.writer;

    // meta block (always present).
    try writer.writeAll("{\n  \"meta\": {\n    \"module\": \"");
    try writeEscaped(writer, doc.meta.module);
    try writer.writeAll("\",\n    \"source\": \"");
    try writeEscaped(writer, doc.meta.source);
    try writer.writeAll("\",\n    \"language\": \"");
    try writeEscaped(writer, doc.meta.language);
    try writer.writeAll("\"\n  }");

    // Early exit when meta is the only field.
    const has_fields = doc.comment != null or doc.detail != null or
        doc.keywords.len > 0 or doc.skills.len > 0 or
        doc.capabilities.len > 0 or doc.hashtags.len > 0 or
        doc.used_by.len > 0 or doc.equivalents.len > 0 or
        doc.members.len > 0;
    if (!has_fields) {
        try writer.writeAll("\n}\n");
        return list_aw.toOwnedSlice();
    }

    // Close meta with comma; remaining fields use writeSep to separate each other.
    try writer.writeAll(",\n");
    var need: bool = false;

    if (doc.comment) |d| {
        try writeSep(writer, &need);
        try writer.writeAll("  ");
        try writeStrField(writer, "comment", d);
    }
    if (doc.detail) |d| {
        try writeSep(writer, &need);
        try writer.writeAll("  ");
        try writeStrField(writer, "detail", d);
    }
    if (doc.keywords.len > 0) {
        try writeSep(writer, &need);
        try writer.writeAll("  ");
        try writeStringArray(writer, "keywords", doc.keywords);
    }
    if (doc.skills.len > 0) {
        try writeSep(writer, &need);
        try writer.writeAll("  \"skills\": [\n");
        for (doc.skills, 0..) |skill, i| {
            try writer.writeAll("    { \"ref\": \"");
            try writeEscaped(writer, skill.ref);
            try writer.writeByte('"');
            if (skill.context) |ctx| {
                try writer.writeAll(", \"context\": \"");
                try writeEscaped(writer, ctx);
                try writer.writeByte('"');
            }
            try writer.writeAll(" }");
            if (i < doc.skills.len - 1) try writer.writeByte(',');
            try writer.writeByte('\n');
        }
        try writer.writeAll("  ]");
    }
    if (doc.capabilities.len > 0) {
        try writeSep(writer, &need);
        try writer.writeAll("  ");
        try writeStringArray(writer, "capabilities", doc.capabilities);
    }
    if (doc.hashtags.len > 0) {
        try writeSep(writer, &need);
        try writer.writeAll("  ");
        try writeStringArray(writer, "hashtags", doc.hashtags);
    }
    if (doc.used_by.len > 0) {
        try writeSep(writer, &need);
        try writer.writeAll("  ");
        try writeStringArray(writer, "used_by", doc.used_by);
    }
    if (doc.equivalents.len > 0) {
        try writeSep(writer, &need);
        try writer.writeAll("  ");
        try writeStringArray(writer, "equivalents", doc.equivalents);
    }
    if (doc.members.len > 0) {
        try writeSep(writer, &need);
        try writer.writeAll("  \"members\": [\n");
        for (doc.members, 0..) |member, i| {
            try writeMember(writer, member, "    ");
            if (i < doc.members.len - 1) try writer.writeByte(',');
            try writer.writeByte('\n');
        }
        try writer.writeAll("  ]");
    }

    try writer.writeAll("\n}\n");
    return list_aw.toOwnedSlice();
}

File: src/guidance/src/sync/json_writer.rs
use guidance_types::{GuidanceDoc, Member};
use serde_json::{json, Value};

fn member_to_json(member: &Member) -> Value {
    let mut obj = json!({
        "type": member_type_str(member.type_name),
        "name": member.name.as_str(),
    });

    if let Some(h) = &member.match_hash {
        obj["match_hash"] = json!(h.as_str());
    }
    if let Some(sig) = &member.signature {
        obj["signature"] = json!(sig.as_str());
    }
    if let Some(ret) = &member.returns {
        obj["returns"] = json!(ret.as_str());
    }
    if let Some(c) = &member.comment {
        obj["comment"] = json!(c.as_str());
    }
    if !member.params.is_empty() {
        let params: Vec<Value> = member
            .params
            .iter()
            .map(|p| {
                let mut po = json!({ "name": p.name.as_str() });
                if let Some(t) = &p.type_name {
                    po["type"] = json!(t.as_str());
                }
                if let Some(d) = &p.default {
                    po["default"] = json!(d.as_str());
                }
                po
            })
            .collect();
        obj["params"] = json!(params);
    }
    if !member.tags.is_empty() {
        let tags: Vec<Value> = member.tags.iter().map(|t| json!(t.as_str())).collect();
        obj["tags"] = json!(tags);
    }
    if member.is_pub {
        obj["is_pub"] = json!(true);
    }
    if !member.members.is_empty() {
        let child: Vec<Value> = member.members.iter().map(member_to_json).collect();
        obj["members"] = json!(child);
    }
    if member.comment_generated {
        obj["comment_generated"] = json!(true);
    }
    if !member.equivalents.is_empty() {
        let eqs: Vec<Value> = member
            .equivalents
            .iter()
            .map(|e| json!(e.as_str()))
            .collect();
        obj["equivalents"] = json!(eqs);
    }

    obj
}

fn member_type_str(t: guidance_types::MemberType) -> &'static str {
    use guidance_types::MemberType::{
        ComptimeBlock, Enum, EnumField, FnDecl, FnPrivate, Method, MethodPrivate, Struct, TestDecl,
        Union,
    };
    match t {
        FnDecl => "fn_decl",
        FnPrivate => "fn_private",
        Struct => "struct",
        Enum => "enum",
        Union => "union",
        EnumField => "enum_field",
        TestDecl => "test_decl",
        ComptimeBlock => "comptime_block",
        Method => "method",
        MethodPrivate => "method_private",
    }
}

pub fn doc_to_json(doc: &GuidanceDoc) -> Value {
    let mut obj = json!({
        "meta": {
            "module": doc.meta.module.as_str(),
            "source": doc.meta.source.as_str(),
            "language": doc.meta.language.as_str(),
        }
    });

    if let Some(c) = &doc.comment {
        obj["comment"] = json!(c.as_str());
    }
    if let Some(d) = &doc.detail {
        obj["detail"] = json!(d.as_str());
    }
    if !doc.keywords.is_empty() {
        let kws: Vec<Value> = doc.keywords.iter().map(|k| json!(k.as_str())).collect();
        obj["keywords"] = json!(kws);
    }
    if !doc.skills.is_empty() {
        let skills: Vec<Value> = doc
            .skills
            .iter()
            .map(|s| {
                let mut so = json!({ "ref": s.ref_path.as_str() });
                if let Some(ctx) = &s.context {
                    so["context"] = json!(ctx.as_str());
                }
                so
            })
            .collect();
        obj["skills"] = json!(skills);
    }
    if !doc.capabilities.is_empty() {
        let caps: Vec<Value> = doc.capabilities.iter().map(|c| json!(c.as_str())).collect();
        obj["capabilities"] = json!(caps);
    }
    if !doc.hashtags.is_empty() {
        let tags: Vec<Value> = doc.hashtags.iter().map(|t| json!(t.as_str())).collect();
        obj["hashtags"] = json!(tags);
    }
    if !doc.used_by.is_empty() {
        let ub: Vec<Value> = doc.used_by.iter().map(|u| json!(u.as_str())).collect();
        obj["used_by"] = json!(ub);
    }
    if !doc.members.is_empty() {
        let members: Vec<Value> = doc.members.iter().map(member_to_json).collect();
        obj["members"] = json!(members);
    }
    if !doc.equivalents.is_empty() {
        let eqs: Vec<Value> = doc.equivalents.iter().map(|e| json!(e.as_str())).collect();
        obj["equivalents"] = json!(eqs);
    }
    if let Some(ce) = &doc.capability_eval {
        obj["capability_eval"] = json!({
            "capability_name": ce.capability_name.as_str(),
            "confidence": ce.confidence,
            "evaluated_at_hash": ce.evaluated_at_hash.as_str(),
        });
    }

    obj
}

pub fn doc_to_json_string(doc: &GuidanceDoc) -> String {
    serde_json::to_string_pretty(&doc_to_json(doc)).unwrap_or_default()
}

#[cfg(test)]
mod tests {
    use super::*;
    use guidance_types::{GuidanceDoc, Member, MemberType, Meta, Param};

    fn make_test_doc() -> GuidanceDoc {
        GuidanceDoc {
            meta: Meta {
                module: "test_module".into(),
                source: "src/test.zig".into(),
                language: "zig".into(),
            },
            comment: Some("A test module.".into()),
            members: vec![Member {
                type_name: MemberType::FnDecl,
                name: "hello".into(),
                signature: Some("fn hello(name: []const u8) -> []const u8".into()),
                params: vec![Param {
                    name: "name".into(),
                    type_name: Some("[]const u8".into()),
                    default: None,
                }],
                returns: Some("[]const u8".into()),
                is_pub: true,
                line: Some(3),
                ..Member::default()
            }],
            ..GuidanceDoc::default()
        }
    }

    #[test]
    fn test_json_round_trip() {
        let doc = make_test_doc();
        let json_str = doc_to_json_string(&doc);
        assert!(json_str.contains("test_module"));
        assert!(json_str.contains("hello"));
        assert!(json_str.contains("fn_decl"));
    }

    #[test]
    fn test_json_sorted_keys() {
        let doc = make_test_doc();
        let json_str = doc_to_json_string(&doc);
        let parsed: serde_json::Value = serde_json::from_str(&json_str).expect("should parse");
        assert_eq!(parsed["meta"]["module"], "test_module");
        assert_eq!(parsed["meta"]["source"], "src/test.zig");
    }
}
